src/AWS/CRT/HTTP/Message.php

```diff
@@ -2,32 +2,94 @@
 
 namespace AWS\CRT\HTTP;
 
-abstract class Message {
+use AWS\CRT\NativeResource;
+
+use AWS\CRT\Internal\Encoding;
+
+abstract class Message extends NativeResource {
+    private $method = "GET";
     private $path = "";
     private $query = [];
     private $headers = [];
 
-    public function __construct($path, $query = [], $headers = []) {
+    public function __construct($method, $path, $query = [], $headers = null) {
+        $this->method = $method;
         $this->path = $path;
         $this->query = $query;
-        $this->headers = $headers;
+        $this->headers = !is_null($headers) ? $headers : new Headers();
```

Comment:
empty() may be better than is_null here-- if someone passes an empty array it would throw exceptions when trying to call methods on that class. This would also be true if they passed in an array with headers instead of a Headers object.
Another option could be making headers into an array again and then set it to always use that as a constructor:
`public function __construct($method, $path, $query = [], $headers = [])`
...
 `$this->headers = new Headers($headers);`

Reply:
It's not an array, it's an object, so `is_null` seems right?

Reply:
What is the ideal behavior if an array of headers is passed?  I would think you'd want to accept the array and turn it into an object.  

Along those same lines, wouldn't you want to accept an empty array of headers and treat it as if it were null?